pkg/consistentfile/file.go

```diff
@@ -0,0 +1,55 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+/*
+Package consistentfile provides a mechanism (on Unix-like platforms) to present a consistent view of a file to separate
+processes even while the file is being written.  This is accomplished by writing a temporary file, syncing to disk, and
+renaming over the destination file name.
+
+Partial/inconsistent reads can occur due to:
+ 1. A process attempting to read the file while it is being written to (both in the case of a new file with a
+    short/incomplete write or in the case of an existing, updated file where new bytes may be written at the beginning
+    but old bytes may still be present after).
+ 2. Concurrent goroutines leading to multiple active writers of the same file.
+
+The above mechanism explicitly protects against (1) as all writes are to a file with a temporary name.
+
+There is no explicit protection against multiple, concurrent goroutines attempting to write the same file. However,
+atomically writing the file should mean only one writer will "win" and a consistent file will be visible.
+
+Note: consistentfile is not implemented for Windows. The Windows codepath returns a normal file.
+*/
+package consistentfile
```

Comment:
Possibly `atomicfile` would be better package name? I see "atomic file" more often mentioned in google for this kind of problems.

Reply:
Sure, I wasn't sure which name was better.  @dcantah (since you were also looking at this) do you have a preference for the name?

Reply:
Sorry I was so focused on perusing windows docs I never reviewed the new package 😭. `atomicfile` makes more sense to me, and maps to what I'd likely grep for first if I went searching to see if a project had some functionality like this already so +1

Reply:
Done.